src/fdc/D88DiskImage.cc

```diff
@@ -0,0 +1,350 @@
+  std::array<uint8_t, 9> reserved;
+  uint8_t writeProtected;
+  uint8_t diskType;
+  int32_t diskSize;
+};
+
+struct D88SectorHeader
+{
+	uint8_t c;
+	uint8_t h;
+	uint8_t r;
+	uint8_t n;
+	uint16_t sectorsPerTrack;
+	uint8_t densityType;
+	uint8_t deleteFlag;
+	uint8_t status;
+	std::array<uint8_t, 5> reserved;
+	uint16_t sectorDataSize;
+};
+
+[[nodiscard]] static /*constexpr*/ bool isValidD88Header(const D88Header& header, const std::array<int32_t, 164>& trackOffsets)
+{
+	if (header.writeProtected != one_of(0x00, 0x10)) {
+		return false;
+	}
+	if (header.diskType != one_of(D88_DISK_TYPE_2DD, D88_DISK_TYPE_1DD)) {
+		return false;
+	}
+
+    // check track offset
+    uint32_t prevTrackOffs = trackOffsets[0];
+	if (prevTrackOffs != sizeof(D88Header) + 4 * D88_TRACK_MAX)
+		return false;
+    for(int32_t i=1; i<160; ++i)
+    {
+        auto trackOffs = trackOffsets[i];
+        auto trackSize = trackOffs - prevTrackOffs;
+        if (trackSize <= 128 || trackSize >= 0x4000)
+            return false;
+		prevTrackOffs = trackOffs;
+    }
+    for(int32_t i=160; i<D88_TRACK_MAX; ++i)
+    {
+        if (trackOffsets[i] != 0)
+            return false;
+    }
+
+	return std::ranges::all_of(header.reserved, [](auto& r) { return r == 0; });
+}
+
+D88DiskImage::D88DiskImage(const Filename& filename, std::shared_ptr<File> file_)
+    : Disk(DiskName(std::move(filename)))
+    , file(std::move(file_))
+	, trackData(8192)
+	, cachedTrackNo(-1)
+{
+    D88Header header;
+    file->seek(0);
+    file->read(std::span{&header, 1});
+    file->read(std::span{&trackOffsets, 1});
+	if (!isValidD88Header(header, trackOffsets)) {
+		throw MSXException("Not a D88 image");
+	}
+
+    diskSize = header.diskSize;
+    singleSided = header.diskType == D88_DISK_TYPE_1DD;
+    writeProtected = header.writeProtected == 0x10;
+
+	// read track 0 first sector
+	D88SectorHeader sectorHeader;
+	file->seek(trackOffsets[0]);
+	file->read(std::span{ &sectorHeader, 1 });
+	sectorsPerTrack = sectorHeader.sectorsPerTrack;
+
+	trackSectors.reserve(32);
+}
+
+bool D88DiskImage::readD88Track(uint8_t track, uint8_t side)
+{
+    int32_t track_ = singleSided ? track : track * 2 + side;
+
+    if (track_ >= 160)
+        return false;
+
+	// check readed track.
+	if (cachedTrackNo == track_)
+		return true;
+	cachedTrackNo = -1;
+
+    auto trackOffs = trackOffsets[track_];
+    auto nextTrackOffs = trackOffsets[track_ + 1];
+    if (nextTrackOffs < 1)
+        nextTrackOffs = diskSize;
+    auto trackSize = nextTrackOffs - trackOffs;
+
+    if (trackOffs==0 || trackSize<1)
+        return false;
+
+	if (trackData.size() < trackSize)
+	{
+		trackData.resize(trackSize);
+	}
+	trackSectors.clear();
+
+	/* Read TrackData */
+    file->seek(trackOffs);
+	file->read(std::span{ trackData });
+
+    uint16_t inTrackOffset = 0;
+    while(inTrackOffset < trackSize)
+    {
+		D88SectorHeader* pHeader = (D88SectorHeader*)(trackData.data() + inTrackOffset);

```

Comment:
We should protect (better) against corrupt D88 disk images:
* It could be that `inTrackOffset < trackSize` (so we enter this `while` loop), but `inTrackOffset + sizeof(D88Sectorheader) > trackSize`. (There's not enough room left for a D88SectorHeader).
* If that's the case then the code below (e.g. `pHeader->sectorDataSize`) will read past the end of the `trackData` buffer.